docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+`internal/` so that we don't end up with all the import version incompatibilities we've described above. However,
+because of the issue with minor version incompatibilities and the need for [unknown field filtering](./adr-020-protobuf-transaction-encoding.md#unknown-field-filtering),
+sticking with the `Keeper` paradigm instead of ADR 033 may be unviable to begin with.
+
+A more performant solution (that could maybe be adapted to work with `Keeper` interfaces) would be to only expose
+getters and setters for generated types and internally store data in memory buffers which could be passed from
+one implementation to another in a zero-copy way.
+
+For example, imagine this protobuf API with only getters and setters is exposed for `MsgSend`:
+```go
+type MsgSend interface {
+	proto.Message
+	GetFromAddress() string
+	GetToAddress() string
+	GetAmount() []v1beta1.Coin
+    SetFromAddress(string)
+    SetToAddress(string)
+    SetAmount([]v1beta1.Coin)
+}
+
+func NewMsgSend() MsgSend { return &msgSendImpl{memoryBuffers: ...} }
+```
+
+Under the hood, `MsgSend` could be implemented based on some raw memory buffers, ex:
+```go
+type msgSendImpl struct {
+	memoryBuffers []interface{}
+}
+
+func (m *msgSendImpl) GetFromAddress() {
+	return m.memoryBuffers[0].(string)
+}
+
+func (m *msgSendImpl) GetToAddress() {
+    return m.memoryBuffers[1].(string)
+}
+```
+
+This approach would have the added benefits of allowing zero-copy message passing to modules written in other languages
+such as Rust. It could also make unknown field filtering in inter-module communication simpler if we require that all
+new fields are added in sequential order, ex. just checking that no field `> 5` is set.
+
+Also, we wouldn't have any issues with state machine breaking code on generated types because all the generated
+code used in the state machine would actually live in the state machine module itself. Depending on how interface
+types and protobuf `Any`s are used in other languages, however, it may still be desirable to take the handler
+approach described in approach A. Either way, types implementing interfaces would still need to be registered
+with an `InterfaceRegistry` as they are now because there would be no way to retrieve them via the global registry.
+
+In order to simplify access to other modules using ADR 033, a public API module (maybe even one
+[remotely generated by Buf](https://docs.buf.build/bsr/remote-generation/go)) could be used by client modules instead
+of requiring to generate all client types internally.
+
+The big downsides of this approach are that it requires big changes to how people use protobuf types and would be a
+substantial rewrite of the protobuf code generator. This new generated code, however, could still be made compatible with
+the [`google.golang.org/protobuf/reflect/protoreflect`](https://pkg.go.dev/google.golang.org/protobuf/reflect/protoreflect)
+API in order to work with all standard golang protobuf tooling.
```

Comment:
Another downsides:
* interface casting
* less friendly code
* more boilerplate code
* performance degradation

Reply:
Why interface casting, performance degradation and boilerplate? The main difference is using setters/getters vs struct fields - I think that's mainly less friendly code, though not really meaningfully more verbose

Reply:
* interface casting, as done in `return m.memoryBuffers[0].(string)` is often considered as something less favorable
* adding all this getters and setters is a boilerplate code
* I think going through protobuf client -> server -> client will involve a bit of performance degradation. Not sure, however, how the bytes will be passed (copied? or only the slice struct will be copied?) and how `memoryBuffers []interface{}` will be reused?